### DISCUSSION TITLE
Can you Render/Draw to a small sub buffer, and then paste it to the main memory space, like a sprite?

### DISCUSSION THREAD

**Discussion by ColinGarySmith** [moononournation/Arduino_GFX#589] [General]
I tried to use the canvas function as per this example but it was crashing the ESP on boot.

I have noticed that Arduino_Canvas *buffer = new Arduino_Canvas(240, 320, display) has input options for output,  outputX and output Y, but i cant find any usage of them.

Definition:

Arduino_Canvas::Arduino_Canvas(
    int16_t w, int16_t h, Arduino_G *output, int16_t output_x, int16_t output_y, uint8_t r)
    : Arduino_GFX(w, h), _output(output), _output_x(output_x), _output_y(output_y)

Even if i can create a frame buffer that i can use as a target for the drawing functions, and then use gfx->drawBitmap to blit it to the main screen that would be fine.

I just need to know how to create a blank unconnected canvas/buffer to act as a draw target for some drawing functions.

Thank you! Btw, your library is actually pretty fantastic, very fast and works very well for what i need it for. Currently working on UI interface elements. I want to be able to do this drawing sprite stuff to allow a "ticker" to be rendered to the screen with zero flicker. All other screen elements are draw as needed, its just the ticker i want to draw into a sub buffer, then blit it. Will stop all flickering.

![image](https://github.com/user-attachments/assets/ecc5ba8d-0836-46e3-aeb7-8fc0e222e083)

My init code:

```

Arduino_ESP32RGBPanel *rgbpanel = new Arduino_ESP32RGBPanel(
    40 /* DE */, 41 /* VSYNC */, 39 /* HSYNC */, 42 /* PCLK */,
    45 /* R0 */, 48 /* R1 */, 47 /* R2 */, 21 /* R3 */, 14 /* R4 */,
    5 /* G0 */, 6 /* G1 */, 7 /* G2 */, 15 /* G3 */, 16 /* G4 */, 4 /* G5 */,
    8 /* B0 */, 3 /* B1 */, 46 /* B2 */, 9 /* B3 */, 1 /* B4 */,
    0 /* hsync_polarity */, 1 /* hsync_front_porch */, 1 /* hsync_pulse_width */, 43 /* hsync_back_porch */,
    0 /* vsync_polarity */, 3 /* vsync_front_porch */, 1 /* vsync_pulse_width */, 12 /* vsync_back_porch */,
    1 /* pclk_active_neg */, 8000000 /* prefer_speed */);
Arduino_RGB_Display *gfx = new Arduino_RGB_Display(
    480 /* width */, 272 /* height */, rgbpanel, 0 /* rotation */, false /* auto_flush */);

Arduino_Canvas *smallBuffer = new Arduino_Canvas(30, 30, gfx);


// in setup.
smallBuffer->begin();



void drawtoBuffer( void )
{
  smallBuffer->startWrite();  
  smallBuffer->drawLine(ORender[i].p0.x, ORender[i].p0.y, ORender[i].p1.x, ORender[i].p1.y, bgC); // Erase old lines
  smallBuffer->endWrite();
  smallBuffer->flush();  // not sure if this is needed?
}


// render to the main screen.

void render(void)
{
  uint16_t *framebuffer = smallBuffer->getFramebuffer();

    // Create an 8-bit array to hold the converted data
    int bufferWidth = 30; // Width of the buffer
    int bufferHeight = 30; // Height of the buffer
    int bufferSize = bufferWidth * bufferHeight * 2; // 2 bytes per pixel (16 bits)
    uint8_t convertedBuffer[bufferSize];

    // Convert the 16-bit framebuffer to an 8-bit array
    for (int i = 0; i < bufferWidth * bufferHeight; i++) {
        convertedBuffer[i * 2] = framebuffer[i] & 0xFF;         // Low byte
        convertedBuffer[i * 2 + 1] = (framebuffer[i] >> 8) & 0xFF; // High byte
      }

    // Specify where to draw the buffer on the main display
    int destX = 100; // Example X position
    int destY = 50;  // Example Y position

    // Use drawBitmap to draw the converted buffer as a sprite onto the main screen.
    gfx->drawBitmap(destX, destY, convertedBuffer, bufferWidth, bufferHeight, 16);
}
```

**Comment by moononournation** [comment]
You may refer to SpriteGif.ino example

**Comment by ColinGarySmith** [comment]
Hi Moon, that example definitely helped, but i dont want to use an offscreen buffer and flush it. 

I have a live screen thats constantly being written to, and my event loop doesn't support a flush, as there is no "end" of the writing.

For the ticker, i want to write it to a "sprite" and then blit that onto the live screen.

**Reply by moononournation** [reply]
should work, you may start revise from that example

**Comment by ColinGarySmith** [comment]
Ok spent a lot of time on this, still fails. I have 4 meg of PSram on this particular device, but as soon as i create a new canvas and draw to it it crashes the ESP32. So there might actually be a hardware issue. Its not an S3 and i think i remember somewhere on your documentation it only works on S3s.
